thread_pool.h

```diff
@@ -25,35 +30,97 @@ class ThreadPool {
     unsigned num_active_threads = 0;
     //FIXME temporary impl
     std::vector<std::thread> threads;
+    bool stoping = false;
+    std::condition_variable event_var;
+    std::mutex event_mutex;
+    std::queue <std::function<void()>>  task_queue;
 
 public:
-    explicit ThreadPool(ThreadPool::Config config = {}) {} //TODO implement
+    explicit ThreadPool(ThreadPool::Config config = {}) {
+        start();
+    } //TODO implement
 
     ~ThreadPool() {
-        //FIXME temporary impl
-        for (auto& th : threads) {
-            th.join();
-        }
+        stop();
         //close active connections
         //reject pending connections
         //send shutdown signal
         //wait for threads destruction
     }
 
-    bool submit_job(std::function<void()>&& job) {
+    auto submit_job(std::function<void()>&& job)->std::future<decltype(job())> 
+    {
+        
         //FIXME temporary impl
-        threads.emplace_back(std::forward<std::function<void()>>(job));
+        //threads.emplace_back(std::forward<std::function<void()>>(job));
         //if accepted
-        return true;
+        if (task_queue::size() < Config::task_queue_size) {
+
+            auto wrapper = std::make_shared<std::packaged_task<decltype(job()) ()>>(std::move(job()));
+
+            {
+                std::unique_lock<std::mutex> lock{ event_mutex };
+                task_queue.emplace([=] {
+                    (*wrapper)();
```

Comment:
This lambda won't do much, since the job has already been called and has blocked the current thread.